src/chunk_index.c

```diff
@@ -573,11 +573,52 @@ typedef struct ChunkIndexDeleteData
 	bool drop_index;
 } ChunkIndexDeleteData;
 
+/*
+ * Lock object.
+ *
+ * In particular, we need to ensure that we lock the table of an index before
+ * locking the index, or run the risk of ending up in a deadlock since the
+ * normal locking order is table first, index second. Since we're not a
+ * concurrent delete, we take a strong lock for this.
+ *
+ * It is also necessary that the parent table is locked first, but we have
+ * already done that at this stage, so it does not need to be done explicitly.
+ */
+static bool
+chunk_lock_object_for_deletion(ObjectAddress *obj)
+{
+	/*
+	 * If we're locking an index, we need to lock the table first. See
+	 * RangeVarCallbackForDropRelation() in tablecmds.c. We can ignore
+	 * partition indexes since we're not using that.
+	 */
+	char relkind = get_rel_relkind(obj->objectId);
+
+	/*
+	 * If we cannot find the object, it might have been concurrently deleted
+	 * (we do not have locks on objects yet).
+	 */
+	if (!relkind)
+		return false;
```

Comment:
Given that this is a char, I find statement  a bit ambiguous. Maybe it is better to explicitly check against the defined values for relkind, such as RELKIND_RELATION? I don't think we expect any other relkinds in this function that tables and indexes, right?

```suggestion
	if (relkind != RELKIND_RELATION && relkind != RELKIND_INDEX)
		return false;
```

or

```suggestion
	if (relkind == '\0')
		return false;
```

Reply:
This is a check if the table exists, so it should be the latter patch in that case. We probably want an assertion as well to ensure that the relkind, if defined, is not incorrect.

Reply:
Looking into the code, it seems like it is possible to have internal dependencies on constraints and sequences as well (and rules, it seems). The only thing we are guaranteed to be able to exclude are partitioned tables.

This is copied from `RangeVarCallbackForDropRelation,` so I think it is safest to handle it the same way as there.